src/modm/communication/amnb/message.hpp

```diff
@@ -0,0 +1,157 @@
+
+#pragma once
+
+#include <modm/architecture/utils.hpp>
+#include <modm/math/utils/crc.hpp>
+#include <memory>
+
+namespace modm::amnb
+{
+
+/// @ingroup modm_communication_amnb
+enum class MessageType: uint8_t
+{
+	Broadcast = 0b00 << 6,	///< Broadcast message without response
+	Request   = 0b01 << 6,	///< Send request to a node
+	Error     = 0b10 << 6,	///< Respond to a request from a node with error code
+	Response  = 0b11 << 6,	///< Respond to a request from a node
+};
+
+/// @ingroup modm_communication_amnb
+struct Message //modm_packed
+{
+public:
+	inline uint8_t address() const          { return header_address; }
+	inline void setAddress(uint8_t address) { header_address = address; }
+
+	inline uint8_t command() const          { return header_command; }
+	inline void setCommand(uint8_t command) { header_command = command; }
+
+	inline uint16_t length() const
+	{
+		const uint8_t len = (header_type_length & LENGTH_MASK);
+		if (len == EXTENDED_LENGTH) return large.data_length;
+		return len;
+	}
+	inline void  setLength(uint16_t length)
+	{
+		if (length <= SMALL_LENGTH) {
+			header_type_length = (header_type_length & ~LENGTH_MASK) | length;
+		}
+		else {
+			header_type_length |= EXTENDED_LENGTH;
+			large.data_length = length;
+			allocateData();
+		}
+	}
+
+	inline MessageType type() const
+	{ return MessageType(header_type_length & ~LENGTH_MASK); }
+	inline void setType(MessageType type)
+	{ header_type_length = (header_type_length & LENGTH_MASK) | uint8_t(type); }
+
+	inline uint8_t* data() const
+	{ return isSmall() ? &small.data : large.data.get(); }
+
+public:
+	inline bool isHeaderValid() const
+	{ return crcHeader() == header_crc; }
+
+	inline bool isDataValid() const
+	{
+		if (not isHeaderValid()) return false;
+		if (dataLength()) return (crcData() == large.data_crc);
+		return true;
+	}
+
+	inline void setValid()
+	{
+		if (not isSmall()) large.data_crc = crcData();
+		header_crc = crcHeader();
+	}
+
+protected:
+	inline uint8_t
+	crcHeader() const
+	{
+		return modm::math::crc8_ccitt(&header_address, headerLength()-1);
+	}
+
+	inline uint16_t
+	crcData() const
+	{
+		return modm::math::crc16_ccitt(large.data.get(), dataLength());
+	}
+
+	inline bool isSmall() const
+	{
+		const uint8_t len = (header_type_length & LENGTH_MASK);
+		return len != EXTENDED_LENGTH;
+	}
+
+	inline uint8_t headerLength() const
+	{
+		const uint8_t len = (header_type_length & LENGTH_MASK);
+		return (len == EXTENDED_LENGTH) ? LARGE_HEADER_SIZE : (SMALL_HEADER_SIZE + len);
+	}
+
+	inline uint8_t dataLength() const
+	{
+		const uint8_t len = (header_type_length & LENGTH_MASK);
+		return (len == EXTENDED_LENGTH) ? large.data_length : 0;
+	}
+
+	inline bool allocateData(uint16_t max_size=0)
+	{
+		if (length() > SMALL_LENGTH) {
+			const size_t size = max_size ? std::min(length(), max_size) : length();
+			large.data.reset(new uint8_t[size]);
+		}
+	}
+
+protected:
+	// NOTE: This arrangement allows this data structure to be 4-byte aligned
```

Comment:
Do you think an [`alignas()` specifier](https://en.cppreference.com/w/cpp/language/alignas) would make sense here to explicitly enforce alignment?

Reply:
Yes!

Reply:
I think ARM EABI requires either 4B or 8B alignment of objects, so I think this isn't useful. On AVRs I'm not sure what the alignment requirements are, could be 2B for pointers?
So not that useful.